pkg/soc/soc.go

```diff
@@ -0,0 +1,243 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package swarm contains most basic and general Swarm concepts.
+package soc
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/ethersphere/bee/pkg/crypto"
+	"github.com/ethersphere/bee/pkg/swarm"
+	bmtlegacy "github.com/ethersphere/bmt/legacy"
+)
+
+const (
+	IdSize        = 32
+	SignatureSize = 65
+	AddressSize   = crypto.AddressSize
+	minUpdateSize = IdSize + SignatureSize + swarm.SpanSize
+)
+
+// Id is a soc identifier
+type Id []byte
+
+// Owner is a wrapper that enforces valid length address of soc owner.
+type Owner struct {
+	address []byte
+}
+
+// NewOwner creates a new Owner.
+func NewOwner(address []byte) (*Owner, error) {
+	if len(address) != AddressSize {
+		return nil, fmt.Errorf("invalid address %x", address)
+	}
+	return &Owner{
+		address: address,
+	}, nil
+}
+
+// Update wraps a single soc update.
+type Update struct {
+	id        Id
+	span      int64
+	payload   []byte
+	signature []byte
+	signer    crypto.Signer
+	owner     *Owner
+}
+
+// NewUpdate creates a new Update from arbitrary soc id and
+// soc byte payload.
+//
+// By default the span of the soc data is set to the length
+// of the payload.
+func NewUpdate(id Id, payload []byte) *Update {
+	return &Update{
+		id:      id,
+		payload: payload,
+		span:    int64(len(payload)),
+	}
+}
+
+// WithSpan is a chainable function that adds an arbitrary
+// span to the chunk data defined for the update.
+func (s *Update) WithSpan(span int64) *Update {
+	s.span = span
+	return s
+}
+
+// WithOwnerAddress provides the possibility of setting the ethereum
+// address for the owner of an update in the absence of a signer.
+func (s *Update) WithOwnerAddress(ownerAddress *Owner) *Update {
```

Comment:
The shorter way of setting multiple stuff could be an `Option` struct with span and owner address fields, but this is good for me as well. Just a suggestion ;)

Reply:
The owner address is set in on of two ways:

* this call
* set the signer, which implicitly sets the address

I felt that having the chainable approach makes it seem more like an optional setting.